### PR TITLE
Audio analysis support and native parallelism

### PR DESCRIPTION
**Pull request by mjs**
* Added support for passing audio files through [audio-analysis](https://github.com/TheCacophonyProject/audio-analysis/)
* Extensive refactoring and clean ups
* Number of workers for each processing step is now set via processing.yaml. There is no longer a need to manage separate systemd services for each processing step.

### CODE REVIEW

File: processing/logs.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import logging
+import multiprocessing
+from logging.handlers import QueueListener, QueueHandler
+
+
+def init_master():
+    q = multiprocessing.Queue()
+
+    handler = logging.StreamHandler()
+    # handler.terminator = "\n\r"
```
Comment (gferraro) [inline processing/logs.py]: remove this

File: processing_TEMPLATE.yaml [hunk 2]
```diff
@@ -38,4 +40,12 @@ tagging:
...
\ No newline at end of file
+    animal_movement: 50
+
+
+audio:
+    convert_workers: 2
+    analysis_workers: 2
+
+    # The command will be called to perform analysis on audio recordings (e.g. Cacophony Index, speech detection)
+    analysis_command: 'docker run -v "{folder}":/io cacophonyproject/audio-analysis /io/"{basename}"'
```
Comment (gferraro) [inline processing_TEMPLATE.yaml]: maybe put the thermal stuff inside it's own nested object as well
```
thermal:
  classify:....
```

File: processing/audio_analysis.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import json
+import logging
```
Comment (gferraro) [inline processing/audio_analysis.py]: is this needed?

File: processing/audio_analysis.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    api = processing.API(conf.api_url)
+    s3 = processing.S3(conf)
+
+    input_extension = mimetypes.guess_extension(recording["rawMimeType"])
+
+    if not input_extension:
+        # Unsupported mimetype. If needed more mimetypes can be added above.
+        logger.error("unsupported mimetype. Not processing")
+        api.report_done(recording, recording["rawFileKey"], recording["rawMimeType"])
```
Comment (gferraro) [inline processing/audio_analysis.py]: will this set anything in the database to represent it failed?

File: processing_TEMPLATE.yaml (hunk 2 above)
Comment (mjs) [reply processing_TEMPLATE.yaml]: Yep, I want to do this but want to do that in a separate PR.

File: processing/audio_analysis.py (hunk 4 above)
Comment (mjs) [reply processing/audio_analysis.py]: Yes there is and there's #7 to actually take advantage of it.

File: processing_TEMPLATE.yaml (hunk 2 above)
Comment (mjs) [reply processing_TEMPLATE.yaml]: https://github.com/TheCacophonyProject/cacophony-processing/issues/54

File: processing/logs.py (hunk 1 above)
Comment (mjs) [reply processing/logs.py]: Oops. That was left over from when I was troubleshooting something.

File: main.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    api = None
+    log_q = None
+
+    def __init__(self, recording_type, processing_state, process_func, num_workers):
+        self.recording_type = recording_type
+        self.processing_state = processing_state
+        self.process_func = process_func
+        self.num_workers = num_workers
+        self.pool = ProcessPool(
+            num_workers, initializer=logs.init_worker, initargs=(self.log_q,)
```
[the remark this answers is not in the document]
Comment (mjs) [reply main.py]: That's to make Python treat the parens as a single item tuple. Without that Python will treat the parens as being for grouping only.

```python
>>> 42 == (42)
True
>>> 42 == (42,)
False
```

File: main.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        )
+        self.in_progress = {}
+
+    def poll(self):
+        self.reap_completed()
+        if len(self.in_progress) >= self.num_workers:
+            return
```
[the remark this answers is not in the document]
Comment (mjs) [reply main.py]: I don't want any jobs being pulled out of the queue without a worker being ready to run the job. This limits the amount of mess to clean up if cacophony-processing is stopped/restarted.

File: main.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            logger.debug(
+                "scheduling %s (%s: %s)",
+                recording["id"],
+                recording["type"],
+                self.processing_state,
+            )
+            future = self.pool.schedule(self.process_func, (recording, self.conf))
+            self.in_progress[recording["id"]] = future
+
+    def reap_completed(self):
```
[the remark this answers is not in the document]
Comment (mjs) [reply main.py]: I'm always a bit wary of the complexity that callbacks introduce but it would probably be fine here. I'm not changing it now :)

**Pull request merged**